### ISSUE TITLE
Making the import process more resilient

### ISSUE THREAD

**Issue by tarebyte** [education/classroom#1253]
Yesterday we had a lot of folks not be able to import repositories because of an issue with the service backing the [Source Imports API](https://developer.github.com/v3/migration/source_imports/)

This tells me that we need to make this process more resilient on our end in order to make it more clear to the end user what is going on.

I propose moving the import process to the background and put the assignments a state of sorts.

These states could be
- `pending`
  - The assignment isn't ready to go
- `ready`
  - The assignment can be worked on by students
- `errored`
  - Something has gone horribly wrong and intervention is probably needed

Unfortunately the only way to know if the import was successful is the [continuously pull from the Source Imports API](https://developer.github.com/v3/migration/source_imports/#get-import-progress). So we'll need create timeouts and retries.

/cc @Kadaaran @d12 for thoughts

**Comment by tarebyte** [comment]
Another thought is that we could add on we could add on to the work that @wilfriedE did with repository imports to make the UI experience even better.

I'm thinking instead of pulling in the foreground we can use action cable and let it listen for the status of the background job.

**Comment by Kadaaran** [comment]
Could we maybe add a time limit when it's `pending` and then update it to `errored` if it takes too long so users know what's happening ?

**Comment by Kadaaran** [comment]
> **TLDR;** This is the new import process design I came up with. We'll back up the process with state machine and secure actions for each state so we won't have issues when Porter is down 💪 

### What exactly is happens
### After a few research here's what I suggest:



With @johndbritton's help , based on talks with @tarebyte and some research I've come to this new design:

<img width="518" alt="screen shot 2018-04-12 at 18 04 02" src="https://user-images.githubusercontent.com/11220823/38989006-9622d186-43d5-11e8-94ca-3c9b2a7207df.png">

I was inspired by the [education-web](https://github.com/github/education-web/blob/master/app/models/discount_request.rb#L33-L65) way of using state machine.
So to help us redesign the process I've thought about using the state-machine gem [Statesman](https://github.com/gocardless/statesman) which seems perfect for what we are trying to do and is still maintained ✨ 


Each state is represented in green.
We're going to have a new class which will keep track of the transition: `ImportStateMachine

**Comment by Kadaaran** [comment]
> TLDR; This is the new import process design I came up with. We'll back up the process with state machine and secure actions for each state so we won't have issues when Porter is down 💪

### What exactly happens

This bug is happening because of the way the action are ordered and *mainly* because we don't take actions when Porter triggers an `Error`.
We add the student to the repo way **before** the code was imported.

https://github.com/education/classroom/blob/4346984e6ca87e555329777b3df0e9ff926ab015/app/models/assignment_repo/creator.rb#L58-L83

### What I suggest

I was inspired by the [education-web's](https://github.com/github/education-web/blob/master/app/models/discount_request.rb#L33-L65) way of using state machine ✋ 
So to help us redesign the process I've been thinking about using the [state-machine gem Statesman](https://github.com/gocardless/statesman) which seems perfect for what we are trying to do and is still maintained ✨

With @johndbritton's help , based on talks with @tarebyte and some research I've come to this new design:
<img width="518" alt="screen shot 2018-04-12 at 18 04 02" src="https://user-images.githubusercontent.com/11220823/38990113-50813c40-43d9-11e8-9896-189054168c5a.png">


#### How does it work ?

Each state is represented in by a green square, which will lead to :  `repo_created`, `started`, `pending`, `complete`,  `student_added`, `access_granted`, `canceled`.
If errors are triggered during the process, we come back to the previous state and remove everything we created.

**__Questions without answers for now__**:

- Do we create a Class without ActiveRecord named `RepoSetup` ?
- Do we create a table `transitions` ?
- We use states when there is a started code, do we use them also for non-starter code repositories ?


Also @tarebyte @johndbritton having your feedback or ideas or comments on it would be so helpful 🙏 

Thanks !

**Comment by BenEmdon** [comment]
> This bug is happening (#1148) because of the way the action are ordered and mainly because we don't take actions (re-run the process or even delete the repo) when Porter triggers an Error.

Im curious about the Porter error:
* What status code is porter sending back? 
* What are the circumstances that cause the failure?

**Comment by BenEmdon** [comment]
I've started tackling this in #1368

**Comment by BenEmdon** [comment]
## Import Resiliency Project Findings
#### The old repo creation flow % failures 
Feb 26th - July 26th
![image](https://user-images.githubusercontent.com/11095731/43938495-c3fae706-9c18-11e8-8185-4594dd8b24a4.png)
![image](https://user-images.githubusercontent.com/11095731/43938547-0b4f62a8-9c19-11e8-86f1-6e74538f5d26.png)


#### The new repo creation flow % failures 
July 28th - Aug 8th
![image](https://user-images.githubusercontent.com/11095731/43938525-f656ceea-9c18-11e8-837a-e05ef5eec005.png)
![image](https://user-images.githubusercontent.com/11095731/43938551-0fe4fabc-9c19-11e8-99d1-28a2b427e915.png)

**Thats a factor of 10 less failures!**

**Comment by d12** [comment]
Nice work!! I'm super happy with the difference we're seeing here.

**Comment by BenEmdon** [comment]
## Group Import resiliency findings
Import resiliency for Group Exercises shipped at the beginning of the school year. Here's the comparison between the old way and the new (smaller is better):
<img width="644" alt="screen shot 2018-10-28 at 4 02 52 pm" src="https://user-images.githubusercontent.com/11095731/47621237-5095a200-dacb-11e8-8554-3344fd89925a.png">
_Data from Jan 1st 2018 to Oct 28th 2018_
  
## Individual assignment resiliency findings
Import resiliency for individual exercises shipped mid summer this year and here's some stats to see its improvement:
<img width="643" alt="screen shot 2018-10-28 at 4 13 11 pm" src="https://user-images.githubusercontent.com/11095731/47621323-6d7ea500-dacc-11e8-868a-d72bb373bea0.png">
_Data from Jan 1st 2018 to Oct 28th 2018_

These numbers are smaller than the earlier reporter numbers since they look over a longer time frame, which mitigates for anomalies (such as the GitHub API going down).

**Comment by BenEmdon** [comment]
A better approach than using the source imports API would be to use GitHub's new template repository feature https://github.blog/2019-06-06-generate-new-repositories-with-repository-templates/

**Comment by d12** [comment]
Closing in favor of many other more up-to-date issues.

https://github.com/education/classroom/issues/1591
https://github.com/education/classroom/issues/2046

Template repos are now enabled as an option for our users, we're working on fixing up some bugs and improving UX and messaging.
